WebdavNode backbone models and collections

PVince81 opened the pull request:
It turns out MKCOL should only be used on create.
The previous patch to this file broke updates with PROPPATCH.

Required to be able to use inner collections for custom groups: https://github.com/owncloud/customgroups/pull/13

@jvillafanez

Commit:
Add Webdav specific models and collections

PVince81 commented:
Uh oh... massive unrelated test failures

PVince81 commented:
Looks like more changes will be needed in the adapter.

It is tricky to make it do a PUT for creating a new object when the id is already provided. Because when the id is already provided, Backbone doesn't consider the object to be a new one so it does an "update" which results in a PROPPATCH. So need a hack to make it do a PUT nevertheless.

But then when changing properties it's also an "update" in which case it must be a PROPPATCH.
Well, PROPPATCH is for changing properties. If the body (if any) would change, it would still need a PUT.

jvillafanez commented:
Is every condition covered? For example, what happens if the method is "create" but the model hasn't an inner collection?

We should consider to split the function into several ones to reduce the code complexity.

The change itself is fine though.

PVince81 commented:
Thanks for the feedback.

I need to rework this piece because it doesn't work quite as I expected.

The main problem is that when a model already has an id, Backbone will automatically make the method "update" even if you call `collection.create({id: 'uri'})`. I need to find a way to bend this to make it work as expected, will likely require more code here.

PVince81 commented:
To understand why we need the different verbs: https://gist.github.com/PVince81/6fce6662d5d37387a8a4c48328f25403

PVince81 commented:
I've now created three types of Backbone objects:

- WebdavNode (backbone model)
    - represents a Webdav node
    - has a special attribute "isNew" to distinguish whether it's new or not as we cannot rely on the "id" attribute (Webdav objects must be created with an already known id (aka uri))
- WebdavCollectionNode (backbone model)
    - extends WebdavNode but represents a Webdav collection
    - contains a sub-collection of type WebdavChildrenCollection: `getChildrenCollection` which contains a list of WebdavNode representing every children node
    - `childrenCollectionClass` is the class used to create the subcollection, default is `WebdavChildrenCollection`
- WebdavChildrenCollection (backbone collection)
    - is a backbone collection and contains subnodes (WebdavNode)
    - contains logic to manage the models special "isNew" attribute

An example usage can be found in the customgroups app: https://github.com/owncloud/customgroups/pull/13

PVince81 commented:
Note that this obsoletes the need for passing `davSync` as a `sync` function. It's all built-in.

Goal is to be able to use these classes in any Webdav-related context in JS.

TODOs:
- [ ] documentation
- [ ] unit tests

@jvillafanez @DeepDiver1975 @felixheidecke please let me know what you think before I start writing tests.

PVince81 commented:
Best is to have a look at https://github.com/owncloud/customgroups/pull/13 which is a good example

PVince81 commented:
Apparently I forgot to `git add` before pushing 😦 
Now it should be fine.

jvillafanez commented on core/js/oc-backbone-webdav.js:
```diff
@@ -287,15 +287,33 @@
+		var isWebdavCollection = !!model.hasInnerCollection;
+
+		// need to override default behavior and decide what to do
+		if (method === 'create') {
```
move these checks to another method and call it here? It seems they're adding unneeded complexity to the davSync method.

PVince81 replied:
yes, will do

jvillafanez commented:
A couple of things I don't understand:

* > has a special attribute "isNew" to distinguish whether it's new or not as we cannot rely on the "id" attribute (Webdav objects must be created with an already known id (aka uri))

    Why do we need to know if it's new or not? I mean, if it's new it will be synced at some point, either implicitly or explicitly. This should be transparent.

* Any reason why the `WebdavCollectionNode` can't handle a list of `WebdavNode`?

PVince81 commented:
> Why do we need to know if it's new or not? I mean, if it's new it will be synced at some point, either implicitly or explicitly. This should be transparent.

Backbone calls the method `isNew()` to decide whether to call the "create" or "update" method and also toggles some logic based on that.

Normally `isNew()` is based on whether the `id` attribute is present. In regular REST, there is no "id" for new objects.

In the case of Webdav, since we're using pre-generated uris as ids, then the id already exists even though the object doesn't exist in the backend. We need to set the "id" attribute to have that uri existing and also to have the MKCOL be done on the URL + URI: (customgroups/$uri) (while in regular REST, one would do a POST on "customgroups/" without id/uri and receive a new id from the server).

Does that make sense ?

PVince81 commented:
> Any reason why the WebdavCollectionNode can't handle a list of WebdavNode?

`WebdavCollectionNode` is a Webdav Collection, but not a Backbone collection. That's the distinction.

A Backbone collection cannot have attributes, it can only contain a list of sub-models. Basically like a glorified array.

But a Webdav collection is both a collection and model at the same time. Since we cannot represent this directly as a single Backbone object, I split both into "WebdavCollectionNode" (the Webdav Collection attributes) and the "WebdavCollectionChildren" (the Webdav Collection array)

PVince81 commented on core/js/oc-backbone-webdav.js:
```diff
@@ -179,10 +179,10 @@
 	}
 
-	function callPropPatch(client, options, model, headers) {
+	function callPropPatch(client, options, model, headers, changed) {
```
Because of https://github.com/owncloud/core/pull/27119/files#diff-052cf83da4708d1c9ce9b1d3574d800cR216

Basically the original logic was to simply read from "model.changed" to find out what attributes changed.

In the normal REST workflow, Backbone would do a POST with **all** attributes and then consider the object successfully created, so it resets `model.changed` to empty. However in the case of Webdav, a `MKCOL` cannot set attributes directly but requires a subsequent PROPPATCH. Here the `MKCOL` plays the same role like `POST` in which Backbone would reset `model.changed` to empty, so we can't use it. So what we do, we re-pass all attributes `model.attributes` as the "changed" param to make sure it will pass all attributes to PROPPATCH.

Hope that makes sense.

So many hoops to jump through with Webdav...

jvillafanez replied:
is resetting the `model.changed` option to true a viable option after sending the `MKCOL` request? I'm not sure if it's too intrusive, but it seems a better option.

PVince81 replied:
I understand your reasoning. But `model.changed` isn't a boolean, it's a hash array of all changed properties as key-values. That's why we need to repass a key-value array here to be set by PROPPATCH.

PVince81 commented on core/js/oc-backbone-webdav.js:
```diff
@@ -287,15 +287,33 @@
-			} else if (model.usePUT || (model.collection && model.collection.usePUT)) {
-				// use PUT instead of PROPPATCH
+		//var isWebdavCollection = (model.get('resourceType').indexOf('{DAV:}collection') >= 0);
+		var isWebdavCollection = !!model.hasInnerCollection;
```
I could replace this with `instanceof WebdavCollection` now

PVince81 commented:
@jvillafanez if you agree with the general approach I intend to write unit tests to solidify these.
When I originally developed this adapter I made the mistake to forget unit tests... I want to at least have these new Webdav* classes tested.

jvillafanez commented:
Ok, I understand the `isNew`. Requesting the server to check if the model is new by doing a propfind or get seems heavier, so I'm fine with the current approach.

PVince81 commented:
Oh look, this extension allows passing properties directly when creating a collection with MKCOL: https://greenbytes.de/tech/webdav/draft-ietf-vcarddav-webdav-mkcol-03.html

MKCOL body:
```xml
<?xml version="1.0" encoding="utf-8" ?>
<D:mkcol xmlns:D="DAV:"
              xmlns:oc="http://owncloud.org/ns">
  <D:set>
    <D:prop>
      <D:resourcetype>
        <D:collection/>
      </D:resourcetype>
      <oc:displayname>display name</oc:displayname>
      <oc:anotherprop>someprop</oc:anotherprop>
    </D:prop>
  </D:set>
</D:mkcol>
```

PVince81 commented:
Well, but if Sabre doesn't support it we'd need to write this ourselves... oh well...

PVince81 commented:
I didn't find any extension for PUT... So it means needs to PUT and then PROPATCH afterwards.

In the past we actually sent the attributes as JSON to PUT.

Maybe have an hybrid Sabre node that returns its properties through PROPFIND but also through GET ?

PVince81 commented:
It turns out, Sabre **does** support extended MKCOL, so we could use that right away.

PVince81 commented:
So, I've added unit tests.

Next up:
- [ ] experiment with extended MKCOL to save additional PROPPATCH
- [ ] add more documentation in the code about how to use WebdavNode, etc

Commit:
Use extended MKCOL in backbone adapter

This makes it possible to pre-set some properties without a subsequent
PROPPATCH call.

PVince81 commented:
@jvillafanez good news ! I was able to get rid of the "changed" logic by bundling the properties directly in an extended MKCOL. See 8011669664e72b3d21749c465c2fec9c79f34a45

- [ ] REQUIRES: This will require an update to davclient.js: https://github.com/evert/davclient.js/pull/25

Commit:
Update davclient.js to get extended mkcol

PVince81 commented:
For now I merged the davclient.js change into my own fork and pointed bower.json to that.
If my PRs get merged eventually we can point bower back.

@jvillafanez please review

jvillafanez commented:
👍
